### ISSUE TITLE
Serial (and vectorised) Reductions + Enhanced Addition Reduction

### ISSUE THREAD

**Issue by damianmoz** [chapel-lang/chapel#24196]
There are times within individual parallel tasks where there is a need for a reduction. Having that reduction work in parallel is counter productive to performance. A serial (and vectorised) version of a reduction would be far more useful.

Also, having a parallel (and serial) reduction called say `++` which implements higher order Kahan summation, i.e. Kahan Babuska Neumaier or Kahan Babuska Klein, would be extremely useful in many situations.  I use such summations far more often than I use a `+` reduction but I can live with a **foreach** for now.

Not urgent. Not stopping anything. Just food for thought for future directions.

**Comment by damianmoz** [comment]
The latter should really return the result as a tuple, an augmented sum, i.e.
```chapel
(x, dx) = ++ reduce a[..]; // where dx << x
```

**Comment by mppf** [comment]
If you are looking for vectorized reductions, you should be able to get that in the future with something like `foreach i in something with (+ reduce total)`. The result will be vectorization & vectorized reduction but not new tasks. However this is not implemented yet.

@stonea has been working in this area.

**Comment by bradcray** [comment]
> If you are looking for vectorized reductions

Ultimately, you should also be able to write something like `var sum = + reduce foreach i in something do i;` but this requires `foreach` expressions which are not yet implemented (see https://github.com/chapel-lang/chapel/issues/19336).

My read of this issue is that it is actually about wanting `+ reduce someExpression` to automatically throttle parallelism such that if `someExpression` is not terribly large, we wouldn't create tasks to compute it, since they could be overkill relative to a simple vectorized loop.  @damianmoz, is that correct?

Damian, with respect for your request for a new flavor of sum reduction, this should be able to be added today (either by us or a sufficiently motivated user) with no language changes if it were to use a name like kbnSum or the like rather than a symbolic name (e.g., `kbnSum reduce.myExpr`.

**Comment by damianmoz** [comment]
@mppf - thanks for the update. That will handle the functionality nicely. Once that is proven, a single keyword such as **vreduce** or **reducev** would make such an operation even more succinct.

@bradcray - in answer to your second paragraph - yes. Actually, the expression can be arbitrarily complex but succinct. The main thing I have seen is that a **reduce** inside a **forall** loop seems to perform woefully, as if it was asking for parallelization resources that have been already exhausted. For my scenarios, which are by definition limited, replacing the reduction by a **for** or **foreach** seems to always yield better performance.

@bradcray - in answer to your third paragraph, I had a look at doing this several times over the last two years. Sadly, all attempts at doing so resulted in my head spinning around and around. The only algorithms I have implemented for these summations are all heavily sequential although vectorizable so my head is likely not yet in the right space. I might attack the problem again in a month or so. I noted this need here because it is related to the first half of the issue.

Thanks for the feedback.

**Comment by stonea** [comment]
> My read of this issue is that it is actually about wanting + reduce someExpression to automatically throttle parallelism such that if someExpression is not terribly large, we wouldn't create tasks to compute it, since they could be overkill relative to a simple vectorized loop

Just to link a related issue: https://github.com/chapel-lang/chapel/issues/22756 (Tom Westerhout showing an example where serial reduction outperforms parallel presumably because of task creation overhead outweighing the benefit of parallelizing it).

**Comment by damianmoz** [comment]
You can also see the problem in the discussion on Discourse about
```
Cache Unfriendly Reduction
```

**Comment by bradcray** [comment]
> The main thing I have seen is that a reduce inside a forall loop seems to perform woefully, as if it was asking for parallelization resources that have been already exhausted.

I don't think it should be doing that (creating nested parallelism), but I wouldn't be surprised if we were still incurring some overhead for dynamically squashing the parallelism in a potentially in a heavyweight way).  It's definitely worth investigating and I've opened up https://github.com/chapel-lang/chapel/issues/24243 to capture that, linking back here, and to Tom's issue.

> Sadly, all attempts at doing so resulted in my head spinning around and around.

That's understandable... Writing a custom reduction in Chapel is not very productive today.  If you could share the code for your serial version of it, though, that would go a long way toward me/us understanding the operation and potentially doing it for you, though.

**Comment by damianmoz** [comment]
Your idea about the overhead of the squashing of the parallelism in the inner loop sounds  like the most sensible explanation.

I will send you my serial version of the higher order Kahan summation later this month, Thanks for the offer.

**Comment by damianmoz** [comment]
Try this.
[kahan.txt](https://github.com/chapel-lang/chapel/files/14022633/kahan.txt)

The correct answer to `sum(a)` is 2.0.  SUM-D is always wrong. SUM-C is sometimes wrong.

Note that `sum(u)` and `sum(v)` are correct.

I do not guarantee the code is bug free or that the parallelization is the smartest.

**Comment by damianmoz** [comment]
The inner serial summation loop should be unrolled eventually for performance. I think there are vectorized versions of this sort of thing which might be a better long term solution. No rush with this.

**Comment by damianmoz** [comment]
That first attempt could potentially be revised.

Given (say) an array `y`which can be partitioned into ranges P[i], the reduction
```
(z, dz) = ++ reduce y;
```
needs to compute in parallel the sums of the individual partitions
```chapel
(x[i], dx[i]) = kbrSum(y[P[i]], (0, 0):y.eltType; // each Sum is done serially
```
and then just return
```chapel
(z, dz) = kbrSum(x, kbrSum(dx, (0, 0):y.eltType));
```

**Comment by damianmoz** [comment]
On an internal review of this Kahan-Babuska-Neumaier summation that having a specialized **reduce** for such compensated summation is just as easily be achieved by a **proc** with little loss of readability. Given the ability to do both parallel KBN and nested serial KBN summation with a **proc**, I agree.

Given that the rest of this issue is now perfectly addressed by Brad's new and more general #24243, I was going to close this issue in the next few days assuming that nobody objects.
